### DISCUSSION TITLE
Solar Eclipse percentage

### DISCUSSION THREAD

**Discussion by bjarkeandersen** [cosinekitty/astronomy#252] [Ideas]
In the method SearchLocalSolarEclipse you get the kind and start, end and peak time for a solar eclipse at a given coordinate. For partial eclipses it would be great to also get the peak percentage. Is that something you plan to add to the library in the nearest future?

**Comment by cosinekitty** [comment]
That is a great idea. For partial and annular eclipses, there is some fraction of the Sun not blocked by the Moon for a given observer. 

However, it's not obvious how to calculate the visible fraction of the Sun, because this is a problem of two overlapping discs. The obvious solution would not be very efficient: essentially pixel counting. Maybe someone else knows how to calculate the area of a disc when another disc of different size overlaps with it?  At any rate, I will think about this, because it would be a handy addition, so long as it doesn't slow down the whole calculation significantly for applications that don't need it.

**Reply by bjarkeandersen** [reply]
If it demands a (too) heavy calculation, you could control it with a bit doCalculatePeakPercentage that defaults to false. This way it will only be calculated when needed.

And by the way - thank you for developing this amazing library!

**Reply by bjarkeandersen** [reply]
I didn't dig into the math, but just tried intuatively to calculate peakPercentage as follows:

`eclipse.peakPercentage = 1 - (shadow.r / shadow.p)`

I tested it at 10 different places and compared with timeanddate.com and it seems to be pretty accurate. Could it be that simple or am I taking too many shortcuts here?

**Reply by hidp123** [reply]
What is `shadow.r` and `shadow.p`?

**Reply by cosinekitty** [reply]
These are internal fields inside the eclipse calculator code. In the case of a local solar eclipse:

`shadow.r` = The distance in kilometers between the observer and the shadow axis (the center of the Moon's shadow).
`shadow.p` = The penumbra radius in kilometers at the plane passing through the observer's location, perpendicular to the shadow axis.

I don't believe the formula above for `peakPercentage` can be correct. For example, even in a total eclipse, it will not say the coverage is 100%, because both `r` and `p` will be positive numbers.

I will need to sit down and think this through, to see if there is an efficient way to calculate coverage. There might be, but it's not obvious.

**Reply by cosinekitty** [reply]
As a follow-up, I realize we would need to be clear about what percentage we are talking about. Is it the percentage of the Sun that is still visible, or the percentage of the Sun that is hidden by the Moon?

**Reply by cosinekitty** [reply]
I found this page about [the area of overlapping circles](https://mathworld.wolfram.com/Circle-CircleIntersection.html). It's a little bit of algebra, but not too crazy. Find where the edges of the circles intersect. Draw sectors from each circle's center to the intersection points. Add the sector areas, then subtract out the triangular part that is inside each sector's chord. The result is the overlapping lens-shaped area. Divide the lens area by the Sun's apparent disc area to get the blocked percentage.

I will work on this over the weekend. I think it's a great thing to add to the local eclipse calculation. I can probably do a similar thing for lunar eclipses: the fraction of the Moon covered by the Earth's umbra. It's a bit more iffy there because of atmospheric effects, but still useful, I imagine.

**Reply by hidp123** [reply]
That's great.

Will the result be geocentric (i.e. same result regardless of observer location) or topocentric? If geocentric, then a topocentric result addition will be good too, as location makes a big difference on eclipses.

**Reply by cosinekitty** [reply]
There are 3 eclipse finders: `SearchGlobalSolarEclipse`, `SearchLocalSolarEclipse`, and `SearchLunarEclipse`.
* For global solar eclipses, I would report the percentage as seen at the point on the Earth where the peak eclipse occurs. The global solar eclipse finder already includes the latitude and longitude of the peak eclipse, so it shouldn't be too hard to use that location as the basis for the percentage calculation.
* For local solar eclipses, I would report the percentage seen by the observer at the requested geographic location.
* For lunar eclipses, I would report the percentage of umbra coverage on the Moon as seen by a geocentric observer, because (a) there is no geographic location specified for this call, and (b) any parallax error would be tiny compared to the blurriness of the actual shadow. (Note that because I also report penumbral lunar eclipses, it is possible for the umbra coverage to be 0%.)

**Reply by bjarkeandersen** [reply]
> As a follow-up, I realize we would need to be clear about what percentage we are talking about. Is it the percentage of the Sun that is still visible, or the percentage of the Sun that is hidden by the Moon?

I guess it doesn't matter as long as it is obvious from the name of the property. The most logic in my head is the percentage of the sun that is covered - i.e. a full eclipse would be 100%

**Reply by cosinekitty** [reply]
> a full eclipse would be 100%

Yes, that's the definition I was thinking of also. Good!

**Reply by bjarkeandersen** [reply]
> These are internal fields inside the eclipse calculator code. In the case of a local solar eclipse:
> 
> `shadow.r` = The distance in kilometers between the observer and the shadow axis (the center of the Moon's shadow). `shadow.p` = The penumbra radius in kilometers at the plane passing through the observer's location, perpendicular to the shadow axis.
> 
> I don't believe the formula above for `peakPercentage` can be correct. For example, even in a total eclipse, it will not say the coverage is 100%, because both `r` and `p` will be positive numbers.
> 
> I will need to sit down and think this through, to see if there is an efficient way to calculate coverage. There might be, but it's not obvious.

You are right about that - at a full eclipse I got values like 99.74% and 99.86% (very close to 100% but obviously not totally correct). It will be interesting to find out the difference between this number at the right number, when you've added the correctly calculated property. I'll take it as an exercise of long forgotten math to find out why they are different.

**Comment by cosinekitty** [comment]
Before adding any new feature to Astronomy Engine, I always make sure I can find some reliable test data. I found this [explanation of eclipse magnitude](https://www.timeanddate.com/eclipse/magnitude.html) on timeanddate.com. It turns out that what we have been talking about, the fraction of apparent area that is blocked (solar eclipse) or darkened by shadow (lunar eclipse) is called *obscuration*.

There is another term *eclipse magnitude* that indicates the fraction of the Sun/Moon's diameter that is blocked/shaded. Eclipse magnitude can sometimes be confused with magnitude, that describes how bright an object is, but is a completely different idea.

I'm open to calculating either/both obscuration (area blockage) and eclipse magnitude (diameter blockage). However, before I do that, I need to find test data. The timeanddate.com site provides eclipse magnitude numbers, so we are covered there.

I thought about trying to convert one to the other, but it won't work. For example, if two different partial solar eclipses block half the diameter the Sun (eclipse magnitude = 0.5), the obscurations can be different, because the Moon can be at a different distance from the Earth, thus having a larger or smaller apparent angular diameter. The same amount of the Sun's diameter can be blocked, but by a different Moon silhouette shape, resulting in a different amount of blocked area. This variability can be surprisingly large because the Moon's orbit is highly eccentric.

Does anyone know of a good source of test data for obscuration numbers?

**Reply by cosinekitty** [reply]
Update: I found part of what I need on Fred Espenak's wonderful site eclipsewise.com. For example, [this page for the October 2022 partial solar eclipse](https://www.eclipsewise.com/solar/SEprime/2001-2100/SE2022Oct25Pprime.html) lists both eclipse magnitude and obscuration numbers for the peak eclipse viewed from the ideal location on the Earth.

It would be better if there was an online tool that allowed entering locations other than the ideal location, for getting a better variety of test data. But that might be wishing for too much, and at any rate, this should be enough for basic validation of solar eclipse calculations.

**Reply by hidp123** [reply]
I found this:
https://tasks.illustrativemathematics.org/content-standards/tasks/1145

Not sure how useful though.

**Reply by hidp123** [reply]
https://www.researchgate.net/publication/315272469_A_study_of_a_solar_eclipse_using_a_photocell

**Reply by hidp123** [reply]
https://www.geogebra.org/m/SnZ7QGTJ

**Reply by hidp123** [reply]
http://www.jgiesen.de/eclipse/index.html

**Reply by hidp123** [reply]
I'm just bulk listing links with eclipse obscuration. You probably would have come across some/all of these. If useful then good, if not then you can happily ignore. You may accordingly delete as well.

**Reply by hidp123** [reply]
https://hemel.waarnemen.com/Computing/eclipse.html

**Reply by hidp123** [reply]
http://xjubier.free.fr/en/site_pages/SolarEclipseCalc_Diagram.html

I have not played with it, but it does have boxes to input observer location. Gives obscuration value as well.

**Reply by hidp123** [reply]
https://www.google.com/search?q=eclipse+%22obscuration%22+calculation

**Reply by cosinekitty** [reply]
Thanks @hidp123 !  The xjubier.free.fr calculator is very helpful. Using that, along with the Espenak data I located, I think I have everything I need now.

I'm starting to think eclipse magnitude is not a very useful number for annular eclipses. It makes sense for a partial eclipse because then it at least correlates with how dark the eclipse is. But I think I would rather just know the obscuration value.

I went back and looked, and sure enough, Fred Espenak does not calculate eclipse magnitude for annular eclipses. In his eclipse prediction tables, for annular eclipses he switches the meaning of the number in the "magnitude" column to be the ratio of the Moon's apparent diameter to the Sun's apparent diameter. It isn't even the fraction of the Sun's diameter anymore. I understand why he does this; it underscores the dubious usefulness of eclipse magnitude.

Obscuration makes so much more sense. For solar eclipses, it tells you what fraction of sunlight is being blocked at a given observation site. For lunar eclipses, it tells you what fraction of the Moon's visible disc is covered by the Earth's umbra.

I see that timeanddate.com gives both umbral coverage and penumbral coverage for lunar eclipses. But I don't see much practical use for penumbral coverage. I have witnessed many lunar eclipses, and I can never discern the penumbra. It's just too subtle for me, and I suspect for most other people also.

So my current inclination is to provide a single obscuration value for all eclipses, and not provide magnitude values. I think we just want to know how dark an eclipse is, in a direct and consistent way for all cases.

**Reply by hidp123** [reply]
http://astro.ukho.gov.uk/eclbin/query_usno.cgi

This reputable site gives obscuration values in %, but not very user friendly.

**Reply by hidp123** [reply]
Another reputable site:
https://aa.usno.navy.mil/data/

Direct link with obscuration %:
https://aa.usno.navy.mil/data/SolarEclipses
https://aa.usno.navy.mil/data/Eclipse2024

Example: https://aa.usno.navy.mil/calculated/eclipse/solar?eclipse=22023&lat=37.76&lon=-122.44&label=San+Francisco%2C+CA&height=0&submit=Get+Data

**Reply by hidp123** [reply]
https://cocalc.com/share/public_paths/4e2e251d8e397d4436d158301ff5f0de9767e81d

Some python code to calculate obscuration.

**Reply by hidp123** [reply]
Stellarium also has some code to calculate obscuration, I'm not sure if/how the value is displayed in the software though:

https://github.com/Stellarium/stellarium/blob/master/src/core/modules/Planet.cpp
(CTRL F, search for "obscuration" on the page)

**Reply by cosinekitty** [reply]
Thank you again, @hidp123!

**Comment by cosinekitty** [comment]
I just released Astronomy Engine v2.1.7, which supports calculation of obscuration for lunar eclipses and solar eclipses. The affected functions are SearchLunarEclipse, SearchGlobalSolarEclipse, and SearchLocalSolarEclipse (along with their respective "Next..." functions). There is a new field `obscuration` in the three data types returned by those functions.
